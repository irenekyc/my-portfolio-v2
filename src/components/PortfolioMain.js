import React from 'react'
import {useSelector} from 'react-redux'
import ProjectCard from './ProjectCard'
import Spinner from './Spinner/Spinner'



const PortfolioMain = ()=>{
    const projects = useSelector(state=> state.data.all)



    return(
        
        <section  className="portfolio-container">
            {projects !== null ? projects.map(p=>{
                return <ProjectCard key={p._id} data={p}/>
            }): <Spinner />}

        </section>
      


    )
}

export default PortfolioMain