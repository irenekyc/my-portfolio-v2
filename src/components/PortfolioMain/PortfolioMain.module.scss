.ikPortfolioMain {
  padding: 1rem 0;
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 1100px;
  margin-right: auto;
  margin-left: auto;
}
