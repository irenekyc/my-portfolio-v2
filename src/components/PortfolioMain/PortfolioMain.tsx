import React, { useState, FunctionComponent, useEffect } from "react";

import ProjectCard from "../ProjectCard/ProjectCard";
import { Project } from "../../typings/Project";

import ModalPortfolio from "../ModalPortfolio";
import styles from "./PortfolioMain.module.scss";
import Pagination from "../Pagination";
import { NUM_PER_PAGE } from "../../constants";

interface PortfolioMainProps {
  projects: Array<Project>;
}

const PortfolioMain: FunctionComponent<PortfolioMainProps> = ({
  projects,
}: PortfolioMainProps) => {
  const [openDetailsModal, setOpenDetailsModal] = useState<Project | null>(
    null
  );
  const [page, setPage] = useState<number>(1);
  const [projectList, setProjectList] = useState<Array<Project>>(projects);
  const [currentPageProjects, setCurrentPageProjects] = useState<
    Array<Project>
  >([]);

  useEffect(() => {
    setProjectList(projects);
    setPage(1);
  }, [projects]);

  const goToPrevPage = () => {
    if (page === 0) return;
    setPage(page - 1);
    // scroll to top
  };

  const goToNextPage = () => {
    setPage(page + 1);
    // scroll to top
  };

  useEffect(() => {
    if (projectList.length === 0) {
      setCurrentPageProjects([]);
    }
    setCurrentPageProjects(
      projectList.slice((page - 1) * NUM_PER_PAGE, page * NUM_PER_PAGE)
    );
  }, [projectList, page]);

  return (
    <>
      <section className={styles.ikPortfolioMain} data-testid="portfolio-main">
        <ModalPortfolio
          show={openDetailsModal !== null}
          projectDetails={openDetailsModal}
          clickClose={() => setOpenDetailsModal(null)}
        />
        {currentPageProjects.map((p: Project) => {
          return (
            <ProjectCard
              key={p.title}
              data={p}
              onClickOpenDetails={(project) => setOpenDetailsModal(project)}
            />
          );
        })}
      </section>
      <Pagination
        hasPrev={page > 1}
        hasNext={Math.ceil(projectList.length / NUM_PER_PAGE) > page}
        goToPrevPage={goToPrevPage}
        goToNextPage={goToNextPage}
      />
    </>
  );
};

export default PortfolioMain;
