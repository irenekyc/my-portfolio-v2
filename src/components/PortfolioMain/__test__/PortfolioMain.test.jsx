import React from "react";
import PortfolioMain from "../index";
import { render, screen } from "@testing-library/react";
import {
  mockProject,
  mockProjectListLong,
} from "../../../helper/__test__/mockdata";

const mockProjectList = [mockProject];

it("Render PortfolioMain Section - number of portfolio card should be zero", () => {
  render(<PortfolioMain projects={[]} />);

  const portfolioMain = screen.getByTestId("portfolio-main");
  expect(portfolioMain).toBeInTheDocument();
  const portfolioCards = screen.queryAllByTestId("portfolio-project-card");
  expect(portfolioCards).toHaveLength(0);
  const pagination = screen.getByTestId("pagination");
  expect(pagination).toBeInTheDocument();
});

it("Render PortfolioMain Section - number of portfolio card should be 1", () => {
  render(<PortfolioMain projects={mockProjectList} />);

  const portfolioMain = screen.getByTestId("portfolio-main");
  expect(portfolioMain).toBeInTheDocument();
  const portfolioCards = screen.queryAllByTestId("portfolio-project-card");
  expect(portfolioCards).toHaveLength(1);
  const pagination = screen.getByTestId("pagination");
  expect(pagination).toBeInTheDocument();
});

it("Render PortfolioMain Section - number of portfolio card should be 9, and has next page", () => {
  render(<PortfolioMain projects={mockProjectListLong} />);

  const portfolioMain = screen.getByTestId("portfolio-main");
  expect(portfolioMain).toBeInTheDocument();
  const portfolioCards = screen.queryAllByTestId("portfolio-project-card");
  expect(portfolioCards).toHaveLength(9);
  const pagination = screen.getByTestId("pagination");
  expect(pagination).toBeInTheDocument();
  const nextButton = screen.getByTestId("pagination-next");
  expect(nextButton).toBeInTheDocument();
});
