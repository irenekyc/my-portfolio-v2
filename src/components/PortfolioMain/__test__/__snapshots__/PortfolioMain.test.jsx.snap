// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render PortfolioMain Section - number of portfolio card should be 9, and has next page 1`] = `
<section
  class="ikPortfolioMain"
  data-testid="portfolio-main"
>
  <div
    class="ikProjectCard__container fadeIn-enter fadeIn-enter-active"
    data-testid="portfolio-project-card"
  >
    <div
      class="ikProjectCard__heading"
    >
      <div
        class="ikProjectCard__heading__tags"
      />
      <div
        class="ikProjectCard__heading__links"
      >
        <a
          data-testid="project-card-github"
          href="https://github.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <i
            class="fab fa-github"
          />
        </a>
      </div>
    </div>
    <div
      class="ikProjectCard__title__div"
    >
      <h2
        class="ikProjectCard__title"
        data-testid="project-card-title"
      >
        title1
      </h2>
    </div>
    <div
      class="ikProjectCard__image"
    >
      <img
        alt=""
        data-testid="project-card-image"
        src="images/image.png"
      />
    </div>
    <div
      class="ikProjectCard__excerpt"
      data-testid="project-card-excerpt"
    >
      <p>
        some excerpt
      </p>
    </div>
    <div>
      <button
        class="ikProjectCard__learnMore"
        data-testid="project-card-learn-more"
        id="title1"
      >
         
        Learn more
      </button>
    </div>
  </div>
  <div
    class="ikProjectCard__container fadeIn-enter fadeIn-enter-active"
    data-testid="portfolio-project-card"
  >
    <div
      class="ikProjectCard__heading"
    >
      <div
        class="ikProjectCard__heading__tags"
      />
      <div
        class="ikProjectCard__heading__links"
      >
        <a
          data-testid="project-card-github"
          href="https://github.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <i
            class="fab fa-github"
          />
        </a>
      </div>
    </div>
    <div
      class="ikProjectCard__title__div"
    >
      <h2
        class="ikProjectCard__title"
        data-testid="project-card-title"
      >
        title2
      </h2>
    </div>
    <div
      class="ikProjectCard__image"
    >
      <img
        alt=""
        data-testid="project-card-image"
        src="images/image.png"
      />
    </div>
    <div
      class="ikProjectCard__excerpt"
      data-testid="project-card-excerpt"
    >
      <p>
        some excerpt
      </p>
    </div>
    <div>
      <button
        class="ikProjectCard__learnMore"
        data-testid="project-card-learn-more"
        id="title2"
      >
         
        Learn more
      </button>
    </div>
  </div>
  <div
    class="ikProjectCard__container fadeIn-enter fadeIn-enter-active"
    data-testid="portfolio-project-card"
  >
    <div
      class="ikProjectCard__heading"
    >
      <div
        class="ikProjectCard__heading__tags"
      />
      <div
        class="ikProjectCard__heading__links"
      >
        <a
          data-testid="project-card-github"
          href="https://github.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <i
            class="fab fa-github"
          />
        </a>
      </div>
    </div>
    <div
      class="ikProjectCard__title__div"
    >
      <h2
        class="ikProjectCard__title"
        data-testid="project-card-title"
      >
        title3
      </h2>
    </div>
    <div
      class="ikProjectCard__image"
    >
      <img
        alt=""
        data-testid="project-card-image"
        src="images/image.png"
      />
    </div>
    <div
      class="ikProjectCard__excerpt"
      data-testid="project-card-excerpt"
    >
      <p>
        some excerpt
      </p>
    </div>
    <div>
      <button
        class="ikProjectCard__learnMore"
        data-testid="project-card-learn-more"
        id="title3"
      >
         
        Learn more
      </button>
    </div>
  </div>
  <div
    class="ikProjectCard__container fadeIn-enter fadeIn-enter-active"
    data-testid="portfolio-project-card"
  >
    <div
      class="ikProjectCard__heading"
    >
      <div
        class="ikProjectCard__heading__tags"
      />
      <div
        class="ikProjectCard__heading__links"
      >
        <a
          data-testid="project-card-github"
          href="https://github.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <i
            class="fab fa-github"
          />
        </a>
      </div>
    </div>
    <div
      class="ikProjectCard__title__div"
    >
      <h2
        class="ikProjectCard__title"
        data-testid="project-card-title"
      >
        title4
      </h2>
    </div>
    <div
      class="ikProjectCard__image"
    >
      <img
        alt=""
        data-testid="project-card-image"
        src="images/image.png"
      />
    </div>
    <div
      class="ikProjectCard__excerpt"
      data-testid="project-card-excerpt"
    >
      <p>
        some excerpt
      </p>
    </div>
    <div>
      <button
        class="ikProjectCard__learnMore"
        data-testid="project-card-learn-more"
        id="title4"
      >
         
        Learn more
      </button>
    </div>
  </div>
  <div
    class="ikProjectCard__container fadeIn-enter fadeIn-enter-active"
    data-testid="portfolio-project-card"
  >
    <div
      class="ikProjectCard__heading"
    >
      <div
        class="ikProjectCard__heading__tags"
      />
      <div
        class="ikProjectCard__heading__links"
      >
        <a
          data-testid="project-card-github"
          href="https://github.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <i
            class="fab fa-github"
          />
        </a>
      </div>
    </div>
    <div
      class="ikProjectCard__title__div"
    >
      <h2
        class="ikProjectCard__title"
        data-testid="project-card-title"
      >
        title5
      </h2>
    </div>
    <div
      class="ikProjectCard__image"
    >
      <img
        alt=""
        data-testid="project-card-image"
        src="images/image.png"
      />
    </div>
    <div
      class="ikProjectCard__excerpt"
      data-testid="project-card-excerpt"
    >
      <p>
        some excerpt
      </p>
    </div>
    <div>
      <button
        class="ikProjectCard__learnMore"
        data-testid="project-card-learn-more"
        id="title5"
      >
         
        Learn more
      </button>
    </div>
  </div>
  <div
    class="ikProjectCard__container fadeIn-enter fadeIn-enter-active"
    data-testid="portfolio-project-card"
  >
    <div
      class="ikProjectCard__heading"
    >
      <div
        class="ikProjectCard__heading__tags"
      />
      <div
        class="ikProjectCard__heading__links"
      >
        <a
          data-testid="project-card-github"
          href="https://github.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <i
            class="fab fa-github"
          />
        </a>
      </div>
    </div>
    <div
      class="ikProjectCard__title__div"
    >
      <h2
        class="ikProjectCard__title"
        data-testid="project-card-title"
      >
        title6
      </h2>
    </div>
    <div
      class="ikProjectCard__image"
    >
      <img
        alt=""
        data-testid="project-card-image"
        src="images/image.png"
      />
    </div>
    <div
      class="ikProjectCard__excerpt"
      data-testid="project-card-excerpt"
    >
      <p>
        some excerpt
      </p>
    </div>
    <div>
      <button
        class="ikProjectCard__learnMore"
        data-testid="project-card-learn-more"
        id="title6"
      >
         
        Learn more
      </button>
    </div>
  </div>
  <div
    class="ikProjectCard__container fadeIn-enter fadeIn-enter-active"
    data-testid="portfolio-project-card"
  >
    <div
      class="ikProjectCard__heading"
    >
      <div
        class="ikProjectCard__heading__tags"
      />
      <div
        class="ikProjectCard__heading__links"
      >
        <a
          data-testid="project-card-github"
          href="https://github.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <i
            class="fab fa-github"
          />
        </a>
      </div>
    </div>
    <div
      class="ikProjectCard__title__div"
    >
      <h2
        class="ikProjectCard__title"
        data-testid="project-card-title"
      >
        title7
      </h2>
    </div>
    <div
      class="ikProjectCard__image"
    >
      <img
        alt=""
        data-testid="project-card-image"
        src="images/image.png"
      />
    </div>
    <div
      class="ikProjectCard__excerpt"
      data-testid="project-card-excerpt"
    >
      <p>
        some excerpt
      </p>
    </div>
    <div>
      <button
        class="ikProjectCard__learnMore"
        data-testid="project-card-learn-more"
        id="title7"
      >
         
        Learn more
      </button>
    </div>
  </div>
  <div
    class="ikProjectCard__container fadeIn-enter fadeIn-enter-active"
    data-testid="portfolio-project-card"
  >
    <div
      class="ikProjectCard__heading"
    >
      <div
        class="ikProjectCard__heading__tags"
      />
      <div
        class="ikProjectCard__heading__links"
      >
        <a
          data-testid="project-card-github"
          href="https://github.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <i
            class="fab fa-github"
          />
        </a>
      </div>
    </div>
    <div
      class="ikProjectCard__title__div"
    >
      <h2
        class="ikProjectCard__title"
        data-testid="project-card-title"
      >
        title8
      </h2>
    </div>
    <div
      class="ikProjectCard__image"
    >
      <img
        alt=""
        data-testid="project-card-image"
        src="images/image.png"
      />
    </div>
    <div
      class="ikProjectCard__excerpt"
      data-testid="project-card-excerpt"
    >
      <p>
        some excerpt
      </p>
    </div>
    <div>
      <button
        class="ikProjectCard__learnMore"
        data-testid="project-card-learn-more"
        id="title8"
      >
         
        Learn more
      </button>
    </div>
  </div>
  <div
    class="ikProjectCard__container fadeIn-enter fadeIn-enter-active"
    data-testid="portfolio-project-card"
  >
    <div
      class="ikProjectCard__heading"
    >
      <div
        class="ikProjectCard__heading__tags"
      />
      <div
        class="ikProjectCard__heading__links"
      >
        <a
          data-testid="project-card-github"
          href="https://github.com"
          rel="noopener noreferrer"
          target="_blank"
        >
          <i
            class="fab fa-github"
          />
        </a>
      </div>
    </div>
    <div
      class="ikProjectCard__title__div"
    >
      <h2
        class="ikProjectCard__title"
        data-testid="project-card-title"
      >
        title9
      </h2>
    </div>
    <div
      class="ikProjectCard__image"
    >
      <img
        alt=""
        data-testid="project-card-image"
        src="images/image.png"
      />
    </div>
    <div
      class="ikProjectCard__excerpt"
      data-testid="project-card-excerpt"
    >
      <p>
        some excerpt
      </p>
    </div>
    <div>
      <button
        class="ikProjectCard__learnMore"
        data-testid="project-card-learn-more"
        id="title9"
      >
         
        Learn more
      </button>
    </div>
  </div>
</section>
`;
